llvm/lib/CodeGen/MachineFunction.cpp

```diff
@@ -699,6 +699,11 @@ bool MachineFunction::needsFrameMoves() const {
          !F.getParent()->debug_compile_units().empty();
 }
 
+/// True if function attribute unsafe-fp-math is true.
+bool MachineFunction::hasUnsafeFPMath() const {
+  return F.getFnAttribute("unsafe-fp-math").getValueAsBool();
```

Comment:
We probably should cache this in a bit in the function instead of requerying the attribute on every use, there are a lot of these. (plus we should stop checking the attributes and move to flags, but that's a bigger ask)

Reply:
Change it to a member variable is ok. Currently `resetTargetOptions` has higher priority than `--enable-unsafe-fp-math`, this results in some unreliable codes in backends and tests, change it to `Options.UnsafeFPMath || F.getFnAttribute("unsafe-fp-math").getValueAsBool()` break some tests.

Reply:
Does `--enable-unsafe-fp-math` currently set `unsafe-fp-math` on all functions?
If it does not, then the change will likely affect performance, as unsafe math will only enable unsafe optimizations in explicitly marked functions only. Ignoring global `Options.UnsafeFPMath` may not be the best idea.

> break some tests.
Can you elaborate on which tests failed and how. It's possible that it's those tests that should be fixed, instead.

Reply:
`--enable-unsafe-fp-math` is always ignored after `resetTargetOptions`, which is called before instruction selection and subtarget info initialization.

Only `CodeGen/NVPTX/fast-math.ll` and `CodeGen/X86/change-unsafe-fp-math.ll` failed.